File: src/main/java/org/jdc/template/util/WebServiceUtil.java
package org.jdc.template.util;

import org.apache.commons.io.FileUtils;

import java.io.File;
import java.io.IOException;

import javax.inject.Inject;
import javax.inject.Singleton;

import okhttp3.ResponseBody;
import retrofit2.Response;
import timber.log.Timber;


@Singleton
public class WebServiceUtil {
    @Inject
    public WebServiceUtil() {
        // Dagger
    }

    public boolean saveResponseToFile(Response<ResponseBody> response, File outputFile) {
        Timber.d("Saving response [%s] to file [%s]...", response.raw().request().url().url().toString(), outputFile.getAbsolutePath());
        boolean success = false;
        try {
            FileUtils.copyInputStreamToFile(response.body().byteStream(), outputFile); // Closes all streams
            success = outputFile.exists();
        } catch (IOException e) {
            Timber.e(e, "Failed to save webservice stream to [%s]", outputFile.getName());

            if (outputFile.exists()) {
                //noinspection ResultOfMethodCallIgnored
                outputFile.delete(); // NOSONAR - just delete if we can but we don't need to check the return value
            }
        }
        return success;
    }
}

File: src/main/kotlin/org/jdc/template/util/WebServiceUtil.kt
package org.jdc.template.util

import okhttp3.ResponseBody
import retrofit2.Response
import timber.log.Timber
import java.io.File
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class WebServiceUtil
@Inject
constructor() {
    fun saveResponseToFile(response: Response<ResponseBody>, outputFile: File): Boolean {
        Timber.d("Saving response [${response.raw().request().url().url()}] to file [${outputFile.absolutePath}]...")
        response.body()?.let {
            outputFile.copyInputStreamToFile(it.byteStream())
        }
        return outputFile.exists()
    }
}
